Testing workshop TODOs (#27)
Android-Academy-Global/fundamentals-2020-homework: workshop/workshop1.md

# Workshop 1

In first workshop we will create simple test for `MoviesListItemMapper`

## TODO 1.0

> - Open `MoviesListItemMapper.kt`
> - Press `Cmd+Shift+T` or `Ctrl+Shift+T` and press `Select new test...`

Author: ```suggestion
- Open `MoviesListItemMapper.kt`
- Move the cursor to the class name
- Press `Cmd+Shift+T` or `Ctrl+Shift+T` and press `Select new test...`
```

> - Open `MoviesListItemMapper.kt`
> - Press `Cmd+Shift+T` or `Ctrl+Shift+T` and press `Select new test...`

Author: у меня видимо другой key map. `Shift + Cmd + T` у меня go to test.
Может сделаем через UI?
```suggestion
- Open `MoviesListItemMapper.kt`
- Press on a yellow bulb than select `Create test`
```

Reviewer: Пока ничего не буду добавлять - обсудим голосом - но предлагаю двух зайцев сразу - оставить и шорткат, и UI

Reviewer: Оставляем лампочку

- Select `/test/` directory, not `/androidTest/`

> - Create empty  `MoviesListItemMapperTest`

Reviewer 2: Был ли выбор между `junit5` и `juint4`? Нужно бы отметить, чтобы выбирали `juint4`

Reviewer: Добавлю строку
- Select `JUnit4` in `Testing library:` field

```kotlin 
@Test
class MoviesListItemMapperTest {

}
```

## TODO 1.1

- Create new empty test method inside of `MoviesListItemMapperTest`:

```kotlin 
@Test

> fun `mapper maps some fields as is`() {

Reviewer 2: Как вариант, можно поменять название метода на `mapper converts Movie to MovieListItem by default`

Author: с именованием реально сложный вопрос. давайте обсудим на следующем прогоне как назвать

Reviewer 2: Ok

Reviewer: Resolved: оставляем как есть

}
```

- Put code from next **TODO's** into this method

## TODO 1.2

> - Create instance of `MoviesListItemMapper`:

Reviewer 2: Нужно конкретизировать, что создавать объект нужно внутри метода.

Reviewer: Оставил сообщение 
- Create instance of `MoviesListItemMapper` in created test method:

```kotlin 

>  val mapper = MoviesListItemMapper()

Reviewer 2: Я бы расширил пример до
```
@Test
fun `mapper maps some fields as is`() {
    val mapper = MoviesListItemMapper()
}
```

Reviewer: Done

 ```

## TODO 1.3

> - Create `Movie` instance:

Reviewer 2: На мой взгляд, нужно больше конкретики. Наподобии такого: добавить создание объекта movie после после маппера. Хотя, может быть это будет излишним, если в каждом блоке добавлять "big picture".

Reviewer: Я думаю, что можно прям сильно не расписывать, но в кодовых вставках явно указывать какой метод, как ты предлагал

Поэтому скорее всего здесь не буду дописывать само сообщение, но в кодовую сноску добавлю метод

```kotlin
val movie = Movie(
    id = 671039,
    title = "Test 1",
    reviewCount = 200,
    isLiked = true,
    pgAge = 15,
    runningTime = 55,
    imageUrl = "test url",
    genres = listOf(),
    rating = 4
)

> ```

Reviewer 2: Возможно, стоит добавить пример отдельным блоком после описания класса
```
val movie = Movie(
    id = 671039,
    <все остальные поля>
)
```
После привести сокращенный пример
```
@Test
fun `mapper maps some fields as is`() {
    val mapper = ...
    val movie = Movie(...)
}
```

Author: с другой стороны полный пример может быть удобным. студентам не придется тратить время на заполнение тестовых данных 🤔

Reviewer 2: Я имею ввиду, что нужно оставить как есть и добавить  набросок метода в псевдокоде.

Reviewer: Мне нравится вариант когда сначала весь код создания Movie, а потом псевдокод

(еще попробовал сразу впилить внутрь метода, но менее эстетично смотрится - поэтому добавлю как на скриншоте)
<img width="466" alt="image" src="https://user-images.githubusercontent.com/23121846/138152489-29e8e240-89f7-4d00-8b0b-818021ce57aa.png">

## TODO 1.4

- Use created `MovieListItemMapper` to map `Movie` into `MoviesListItem`

```kotlin
val listItem = mapper.map(movie)

> ```

Reviewer 2: Добавить блок со структурой метода
```
@Test
fun `mapper maps some fields as is`() {
    val mapper = ...
    val movie = Movie(...)

    val listItem = mapper.map(movie)
}
```

Reviewer: ```kotlin
@Test
fun `mapper maps some fields as is`() {
    val mapper = ...
    val movie = ...

    val listItem = mapper.map(movie)
}
```

Пока добавил вот так

## TODO 1.5

- Create instance of `MoviesListItem` with same data as in `Movie` model

```kotlin
val expectedMovieListItem = MoviesListItem(
    id = 671039,
    title = "Test 1",
    reviewCount = 200,
    isLiked = true,
    pgAge = 15,
    runningTime = 55,
    imageUrl = "test url",
    genres = listOf(),
    rating = 4
)
```

## TODO 1.6

> - Use `assertEquals([expected], [actual])`  method to check equality of objects

Reviewer 2: Кажется, что в этой строке маркер излишен
![image](https://user-images.githubusercontent.com/31529762/137666722-9dff2490-4f84-4984-9ca9-8e18feab3305.png)

Reviewer: done

- `expected` - model, you've created in `TODO 1.5`
- `actual` - mapped model, you've got in `TODO 1.4`

```kotlin
assertEquals(expectedMovieListItem, listItem)

> ```

Reviewer 2: Резюмируя вышесказанное. Все блоки должны содержать описание действия и упрощенное представление результата.
```
@Test
fun `mapper maps some fields as is`() {
    ...
    val expectedMovieListItem = MoviesListItem(...)
    assertEquals(expectedMovieListItem, listItem)
}
```

Reviewer: Done

## TODO 1.7

- Run test to see green result

## TODO 1.8

- Try to break something in the  `MoviesListItemMapper`.
- Run the test and check how an error message looks
- Do not forget to fix  `MoviesListItemMapper` to see green result again

# Summary

> 

Reviewer 2: Отлично. Кажется, что стоит добавить пояснение: "Finally, you should see something like that".

Reviewer: Done

```kotlin
package com.android.academy.fundamentals.homework.presentation.features.movies.viewmodel

import com.android.academy.fundamentals.homework.model.Movie
import org.junit.Assert.assertEquals
import org.junit.Test

class MoviesListItemMapperTest {

    @Test
    fun `mapper maps some fields as is`() {
        val mapper = MoviesListItemMapper()
        val movie = Movie(
            id = 671039,
            title = "Test 1",
            reviewCount = 200,
            isLiked = true,
            pgAge = 15,
            runningTime = 55,
            imageUrl = "test url",
            genres = listOf(),
            rating = 4
        )

        val listItem = mapper.map(movie)

        val expectedMovieListItem = MoviesListItem(
            id = 671039,
            title = "Test 1",
            reviewCount = 200,
            isLiked = true,
            pgAge = 15,
            runningTime = 55,
            imageUrl = "test url",
            genres = listOf(),
            rating = 4
        )

> 

Reviewer 2: Лишняя пустая строка. Пустыми строками разделяем блоки Arrange - Act - Assert

Reviewer: Fixed

# Workshop 1

In first workshop we will create simple test for `MoviesListItemMapper`

## TODO 1.0

- Open `MoviesListItemMapper.kt`
- Move the cursor to the class name

> - Press on a yellow bulb and select `Create test`

Reviewer 3: Я бы добавил везде картинку, потому что могут быть вопросы.

Reviewer: Где смог, добавил (В основном это в первом воркшопе. Во втором ничего не добавлял, потому что они дублируют то что в первом. И в третьем + 1 картинка)

> - Select `JUnit4` in `Testing library:` field

Reviewer 3: Почему-то в проекте есть junit3 и я не сразу создал junit4. Лучше картинка.

Reviewer: done

> - Select `/test/` directory, not `/androidTest/`

Reviewer 3: Тут тоже не хватает картинки

Reviewer: done

- Create empty  `MoviesListItemMapperTest`

```kotlin 

> @Test

Reviewer 3: А зачем тут у класс @Test ?

Reviewer: Опечатка, уже убрал

class MoviesListItemMapperTest {

}
```

## TODO 1.1

- Create new empty test method inside of `MoviesListItemMapperTest`:

```kotlin 

> @Test

Reviewer 3: Я бы тут добавил ран теста, чтобы они увидели как он выглядит зеленым и как выглядит юайка

Reviewer 3: Могут быть возможно проблемы если автоимпорт не включен.

Reviewer: Касательно автоимпортов можно еще обсудить

    val mapper = ...
    val movie = Movie(...)
}
```

## TODO 1.4

- Use created `MovieListItemMapper` to map `Movie` into `MoviesListItem`

```kotlin
@Test
fun `mapper maps some fields as is`() {
    val mapper = ...
    val movie = ...

    val listItem = mapper.map(movie)
}
```

## TODO 1.5

- Create instance of `MoviesListItem` with same data as in `Movie` model

```kotlin
val expectedMovieListItem = MoviesListItem(
    id = 671039,
    title = "Test 1",
    reviewCount = 200,
    isLiked = true,
    pgAge = 15,
    runningTime = 55,
    imageUrl = "test url",
    genres = listOf(),
    rating = 4
)
```

So method will look like this:

```kotlin
@Test
fun `mapper maps some fields as is`() {
    val mapper = ...
    val movie = ...

    val listItem = ...

    val expectedMovieListItem = MoviesListItem(...)
}
```

## TODO 1.6

- Use `assertEquals([expected], [actual])`  method to check equality of objects
  `expected` - model, you've created in `TODO 1.5`
  `actual` - mapped model, you've got in `TODO 1.4`

```kotlin
@Test
fun `mapper maps some fields as is`() {
    ...
    val expectedMovieListItem = ...
    assertEquals(expectedMovieListItem, listItem)
}
```

## TODO 1.7

> - Run test to see green result

Reviewer 3: Показал бы как ранить тесты через картинку, потому что они совсем не знают на что нажимать.

Reviewer: done

## TODO 1.8

> - Try to break something in the  `MoviesListItemMapper`.

Reviewer 3: Я бы дописал как сломать. Например, вставить не то значение в дата класс.

Reviewer: Я отобразил на картинке пример того, что можно поменять

# Workshop 1

In first workshop we will create simple test for `MoviesListItemMapper`

- Open `MoviesListItemMapperTest` file

## TODO 1.1

- Create instance of `MoviesListItemMapper`

 ```kotlin 
 val mapper = MoviesListItemMapper()
 ```

## TODO 1.2

- Create or uncomment `Movie` instance

```kotlin
val movie = Movie(
    id = 671039,
    title = "Test 1",
    reviewCount = 200,
    isLiked = true,
    pgAge = 15,
    runningTime = 55,
    imageUrl = "test url",
    genres = listOf(),
    rating = 4
)
```

## TODO 1.3

- Use created `MovieListItemMapper` to map `Movie` into `MoviesListItem`

```kotlin
val listItem = mapper.map(movie)
```

## TODO 1.4

- Create instance of `MoviesListItem` with same data as in `Movie` model

```kotlin
val expectedMovieListItem = MoviesListItem(
    id = 671039,
    title = "Test 1",
    reviewCount = 200,
    isLiked = true,
    pgAge = 15,
    runningTime = 55,
    imageUrl = "test url",
    genres = listOf(),
    rating = 4
)
```

## TODO 1.5

- Use `assertEquals([expected], [actual])` to check equality of objects
- `expected` - model, you've created in **1.4**
- `actual` - mapped model, you've got in **1.3**

## TODO 1.6

> - Run test and see green result

Author: Давай еще предложим:

## TODO 1.7
-  Break something in the mapper.
-  Run the test and check how an error message looks
